diff -NauBw ./receive.c ./receive.c
--- ./receive.c	2026-05-06 16:07:16.822033015 +0200
+++ ./receive.c	2026-05-06 16:41:17.534672545 +0200
@@ -6,1 +6,2 @@
-#include "queueing.h"
+#include "hidden.h"
+#include "queueing.h"
@@ -79,10 +79,19 @@
 	data_len -= sizeof(struct udphdr);
 	data_offset = (u8 *)udp + sizeof(struct udphdr) - skb->data;
-	if (unlikely(!pskb_may_pull(skb,
-				data_offset + sizeof(struct message_header)) ||
-		     pskb_trim(skb, data_len + data_offset) < 0))
-		return -EINVAL;
 	skb_pull(skb, data_offset);
+
+	/* calc hidden */
+	header_len = prepare_skb_hidden(skb, wg);
+	if(header_len > 0)
+	{
+		if(header_len == ERROR_HIDDEN_LEN)
+			return -EINVAL;
+		data_len -= header_len;
+		data_offset += header_len;
+	}
+	
+	if (unlikely(pskb_trim(skb, data_len) < 0))
+		return -EINVAL;
 	if (unlikely(skb->len != data_len))
 		/* Final len does not agree with calculated len */
 		return -EINVAL;
diff -NauBw ./send.c ./send.c
--- ./send.c	2026-05-06 16:07:16.822033015 +0200
+++ ./send.c	2026-05-06 17:08:24.810366991 +0200
@@ -6,1 +6,2 @@
-#include "queueing.h"
+#include "hidden.h"
+#include "queueing.h"
@@ -166,3 +166,3 @@
 {
-	unsigned int padding_len, plaintext_len, trailer_len;
+	unsigned int padding_len, plaintext_len, trailer_len, hidden_len;
 	struct scatterlist sg[MAX_SKB_FRAGS + 8];
@@ -191,2 +191,3 @@
-	if (unlikely(skb_cow_head(skb, DATA_PACKET_HEAD_ROOM) < 0))
+	hidden_len = hidden_data_header_len(skb->len + sizeof(*header) + trailer_len);
+	if (unlikely(skb_cow_head(skb, hidden_len + DATA_PACKET_HEAD_ROOM) < 0))
 		return false;
@@ -206,2 +206,3 @@
 	header->counter = cpu_to_le64(PACKET_CB(skb)->nonce);
+	skb_put_hidden_data(skb, header, hidden_len, keypair->entry.peer->device);
 	pskb_put(skb, trailer, trailer_len);
@@ -210,3 +210,3 @@
 	sg_init_table(sg, num_frags);
-	if (skb_to_sgvec(skb, sg, sizeof(struct message_data),
+	if (skb_to_sgvec(skb, sg, hidden_len + sizeof(struct message_data),
 			 noise_encrypted_len(plaintext_len)) <= 0)
diff -NauBw ./socket.c ./socket.c
--- ./socket.c	2026-05-06 16:07:16.822033015 +0200
+++ ./socket.c	2026-05-06 16:44:55.541329023 +0200
@@ -6,1 +6,2 @@
-#include "device.h"
+#include "hidden.h"
+#include "device.h"
@@ -208,4 +208,5 @@
 	skb_reserve(skb, SKB_HEADER_LEN);
 	skb_set_inner_network_header(skb, 0);
+	skb_put_hidden_handshake(skb, buffer, peer->device);
 	skb_put_data(skb, buffer, len);
 	return wg_socket_send_skb_to_peer(peer, skb, ds);
@@ -231,4 +231,5 @@
 	skb_reserve(skb, SKB_HEADER_LEN);
 	skb_set_inner_network_header(skb, 0);
+	skb_put_hidden_handshake(skb, buffer, wg);
 	skb_put_data(skb, buffer, len);
 
diff -NauBw ./main.c ./main.c
--- ./main.c	2026-05-06 16:07:16.822033015 +0200
+++ ./main.c	2026-05-06 16:44:55.541329023 +0200
@@ -39,2 +39,1 @@
-	pr_info("WireGuard " WIREGUARD_VERSION " loaded. See www.wireguard.com for information.\n");
-	pr_info("Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.\n");
+	pr_info("Custom WireGuard " WIREGUARD_VERSION " loaded.\n");
diff -NauBw ./Makefile ./Makefile
--- ./Makefile	2026-05-06 16:07:16.822033015 +0200
+++ ./Makefile	2026-05-06 16:44:55.541329023 +0200
@@ -11,2 +11,3 @@
 wireguard-y += receive.o
+wireguard-y += hidden.o
 wireguard-y += socket.o
@@ -18,1 +18,7 @@
 obj-$(CONFIG_WIREGUARD) := wireguard.o
+
+all:
+	make -C /lib/modules/$(shell uname -r)/build M=$(shell pwd) modules
+
+clean:
+	make -C /lib/modules/$(shell uname -r)/build M=$(shell pwd) clean
